### DISCUSSION TITLE
Note をTechArticle スコア に追加

### DISCUSSION THREAD

**Discussion by nunukim** [lapras-inc/lapras-user-community#27] [スコアについてのご意見・議論]
# 概要

現在、Note 記事はクロール対象だがスコアの計算に考慮されていないので、これを入れることについて検討したいです。

* そもそも入れるべきか？
* 入れるならどうするか？

# 論点

### どう算入するか？

* Note 記事は現在クロールしていて、記事ごとのStar 数も取れているので、今のTech Article にそのまま追加することができそう。
* 現状のロジックにそのまま加える場合、「Qiita, Zenn, Note のうち、Star(Stock)数の上位3件の対数の積」になる。
    * そもそも、Note のstarはQiita, Zenn のstar(stock)と並列で並べてよいものなのか？
    * サービスが増えてきたので、上位3つしか使っていない問題がさらに大きくなってしまわないか？ (関連 → #24)

### 「技術記事」をどう判定するか？

技術力スコアに入れるのであれば、対象は技術記事に絞るべき。どのように技術記事を判定するか

* タイトルで判定する
    * タイトルから技術記事か否かを判定する二値分類器を作る
    * 簡易的に正規表現や固有表現抽出で技術系のキーワードが含まれるかどうかで判定する。
* トピックで絞る（例えば [IT](https://note.com/topic/it) のみに絞る）

**Comment by Haruki-Oni** [comment]
noteで結構書いているので、noteがスコア対象に入ると嬉しい。

**Comment by yuichkun** [comment]
# 社内検討進捗

## 現状取れているデータの調査

**記事のタイトル・like数・ハッシュタグ**等がデータとして取れている

LAPRAS上の画面でいうと、↓の箇所
![Screenshot 2025-01-27 at 17 25 12](https://github.com/user-attachments/assets/18717601-6608-4cd7-931f-61b097f80068)

## note記事が「技術記事」であると判定する方法検討

### 1. noteのカテゴリーを利用する方法

> [!IMPORTANT]  
> 結論: クロールが難しく、かつnoteユーザーが自分で明示的に設定できないので、諦めることになった

<details>
<summary>詳細</summary>

[noteの公式情報](https://www.help-note.com/hc/ja/articles/40819201911577-%E8%87%AA%E5%88%86%E3%81%AE%E8%A8%98%E4%BA%8B%E3%81%8C%E3%81%A9%E3%81%AE%E3%82%AB%E3%83%86%E3%82%B4%E3%83%AA%E3%81%AB%E5%88%86%E9%A1%9E%E3%81%95%E3%82%8C%E3%81%A6%E3%81%84%E3%82%8B%E3%81%8B%E7%A2%BA%E8%AA%8D%E3%81%99%E3%82%8B%E6%96%B9%E6%B3%95%E3%82%84-%E5%88%86%E9%A1%9E%E3%81%95%E3%82%8C%E3%82%8B%E5%9F%BA%E6%BA%96%E3%82%92%E6%95%99%E3%81%88%E3%81%A6%E3%81%8F%E3%81%A0%E3%81%95%E3%81%84)によると、

> カテゴリの分類は自動で行われており、記事がピックアップされる条件や表示順、表示されるタイミングなどにつきましては、システムの仕組みに関わるため、詳細は公開していません。

とある。

また、noteの記事ページをみても、「その記事がどのカテゴリに属しているか」を明示的に確認することはできなさそう。
</details>


### 2. 記事につけられているハッシュタグを利用する

ハッシュタグについては、ユーザーが自分の記事に対して、任意のハッシュタグを任意の数つけられる。

note上ではこのように表示されて確認できる。
例: https://note.com/yuichi_yogo/n/n066a3f997a94
![Screenshot 2025-01-27 at 17 33 52](https://github.com/user-attachments/assets/d543b641-dcf2-4c00-bd28-3e0e1a34f4d9)

## どのようなハッシュタグを「技術記事」として判定するか

> [!IMPORTANT]  
>  note記事のハッシュタグに基づく「技術記事」としての判定方法は、今まさに社内検討中なので、LAPRASとしては**コミュニティの意見を広く募集したいです** 🙇‍♂️

### 考慮する点としてこれまでに話題にあがっていること

- アプローチ案（ざっくり）
    - 技術記事として判定するタグ一覧を用意し、メンテしていく
        - 考慮すべき点
            - 初期データをどのように作るか    
                - LAPRASサービス上では、一部技術用語のマスターデータを持っている
                    - 「Tag」や「経験技術」など
            - ハッシュタグは固有名詞そのままで使われないこともある
                - 例: `#ruby` だけではなく `#rubykaigi` を検知できるようにするためには?
                    - 形態素解析等を行うかどうか
    - LLM等で技術記事かどうか判定
        - メリット
            - 内容にまで踏み込んで判定できる
        - 考慮すべき点
            - 判定結果を毎回同じにできるかどうか
            - サービス名・フレームワーク名などの固有名詞をちゃんと認識できるかどうか
            - オンデマンドで判定するか、一度だけ判定して結果を保存しておくか
                - 前者の場合、計算コスト・スピードがどうなるか
                - 後者の場合、ユーザーが記事を更新する場合のライフサイクルがどうなるか
- 正確性/検出できる広さについて
    - どのようなアプローチを取ったとしても、正確性 ↔️検出できる広さの間にはトレードオフの関係がある
        - より正確に厳密にハッシュタグを選定するほど、**偽陽性⤵ & 偽陰性⤴**  になる
            - = 技術記事ではないのに技術記事としてカウントされるケースは少ないが、**本当は技術記事なのに技術記事としてカウントされない**ケースも増える
        - より広くハッシュタグを選ぶほど、**偽陽性⤴ & 偽陰性⤵**  になる
            - = 技術記事として検出できる数は増えるが、**本当は技術記事ではないものも拾ってしまう可能性が高く**なる
- 運用面
    - たとえば以下のような運用が可能そうかどうか
        - 「〇〇という記事が、技術記事なのにカウントされていない」というお問い合わせ
            - 技術記事としてカウントされるタグ一覧が、ユーザーにわかりやすい形で用意できるかどうか
        - 「〇〇というタグも技術記事としてカウントされるべきではないか」というお問い合わせ
            - LAPRASの担当者がシステム管理画面で手動で更新追加するフローにする / データベース上のマスターデータの更新をする形になるか、など
- データのライフサイクル
    - ハッシュタグを更新した場合、どのタイミングでデータベース上のハッシュタグが更新され、どのタイミングでスコアの再計算が行われるか、など
